### PR TITLE
ARROW-5248: [Python] support zoneinfo / dateutil timezones

### PR DESCRIPTION
**Pull request by AlenkaF**
This PR tries to add support for `zoneinfo` and `dateutil` timezones. 

- [x] Add support for the `zoneinfo`
- [x] Add tests for `zoneinfo` support
- [x] Add support for `dateutil`
- [x] Add tests for `dateutil` support

### CODE REVIEW

File: cpp/src/arrow/python/datetime.cc [hunk 1]
```diff
@@ -470,6 +474,26 @@ Result<std::string> TzinfoToString(PyObject* tzinfo) {
...
+  }
+
+  // try to look up _filename attribute
+  if (PyObject_HasAttrString(tzinfo, "_filename")) {
+    OwnedRef _filename(PyObject_GetAttrString(tzinfo, "_filename"));
+    RETURN_IF_PYERROR();
+    std::string result;
+    RETURN_NOT_OK(internal::PyUnicode_AsStdString(_filename.obj(), &result));
+    std::size_t pos = result.find("zoneinfo/");
+    if (pos > 0) {return result.substr (pos+9);}
```
Comment (AlenkaF) [inline cpp/src/arrow/python/datetime.cc:493]: I am not sure this is a way to do things in C++ part of Arrow. So I will add a note to clarify. `Dateutil` package currently does not have a method to extract POSIX name of the timezone. It is possible to use `_filename` but the output can be `'/usr/share/zoneinfo/Europe/Brussels'` or `'Europe/Brussels'` as in example:

```
>>> import dateutil.tz
>>> tz = dateutil.tz.gettz('Europe/Brussels')
>>> tz
tzfile('/usr/share/zoneinfo/Europe/Brussels')
>>> tz._filename
'/usr/share/zoneinfo/Europe/Brussels'
```
see  https://github.com/dateutil/dateutil/issues/76#issuecomment-91226288.

For this reason I am selecting a substring to get the name from the `tz._filename.` Please let me know if this is not a good way to go.

File: cpp/src/arrow/python/datetime.cc [hunk 2]
```diff
@@ -470,6 +474,28 @@ Result<std::string> TzinfoToString(PyObject* tzinfo) {
     return result;
   }
 
+  // try to look up key attribute
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/python/datetime.cc:477]: Maybe add a comment here that this check is explicitly for zoneinfo objects, and below the same for dateutil

**Review by jorisvandenbossche — COMMENTED** [review]
This is looking good!

From talking about it:

- Dateutil is optional dependency, and zoneinfo only available from Python 3.9 onwards, so in C++ we (for now) can remove the imports
- The tests will need to add a skip if the package is not available (pytest has a `importorskip` helper that can be used for this)

**Comment by AlenkaF** [comment]
Thanks for reviewing Joris! Currently:

- imports are removed
- test is split and skipped in case `zoneinfo` is not available
- comments for the check in case of `zoneinfo` and `dateutil` objects are added

I would leave the handling of `zoneinfo` and `dateutil` objects in `TzinfoToString` as is (looking up 'key'  and '_filename' attribute) and plan to restructure (to check for an instance of the object) together with `pytz` change for https://issues.apache.org/jira/browse/ARROW-15580.

There is a CI error with appveyor: `43 - arrow-dataset-file-csv-test (Failed)` in ctest. Don't know what would cause that -will look into it a bit more.

**Comment by jorisvandenbossche** [comment]
> There is a CI error with appveyor: `43 - arrow-dataset-file-csv-test (Failed)` in ctest. Don't know what would cause that -will look into it a bit more.

That looks like an unrelated error. I restarted Appveyor to see if that fixes it.

File: cpp/src/arrow/python/datetime.cc [hunk 3]
```diff
@@ -427,6 +427,8 @@ Result<PyObject*> StringToTzinfo(const std::string& tz) {
 Result<std::string> TzinfoToString(PyObject* tzinfo) {
   OwnedRef module_pytz;        // import pytz
   OwnedRef module_datetime;    // import datetime
+  OwnedRef module_zoneinfo;    // import zoneinfo
+  OwnedRef module_dateutil;    // import dateutil
```
[the remark this answers is not in the document]
Comment (AlenkaF) [reply cpp/src/arrow/python/datetime.cc:431]: Oh, yes, will do!

**Comment by AlenkaF** [comment]
@jorisvandenbossche it looks like the error on Travis isn't connected, so this PR should be ready.

File: python/pyarrow/tests/test_types.py [hunk 4]
```diff
@@ -303,6 +303,24 @@ def test_tzinfo_to_string(tz, expected):
     assert pa.lib.tzinfo_to_string(tz) == expected
 
 
+def test_dateutil_tzinfo_to_string():
+    import dateutil.tz
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_types.py:307]:
```suggestion
    pytest.importorskip("dateutil")
    import dateutil.tz
```

**Comment by AlenkaF** [comment]
@jorisvandenbossche unrelated Ci issue again, I think ...

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request closed without merging**